internal/packager/fork.go

```diff
@@ -0,0 +1,122 @@
+package packager
+
+import (
+    "archive/tar"
+    "io"
+    "io/ioutil"
+    "os"
+    "path/filepath"
+    "strings"
+
+    "github.com/docker/app/internal"
+    "github.com/docker/app/internal/types"
+
+    "github.com/pkg/errors"
+    log "github.com/sirupsen/logrus"
+    yaml "gopkg.in/yaml.v2"
+)
+
+// Fork pulls an application and creates a local fork for the user to modify
+func Fork(originName, forkName, outputDir string, maintainers []string) error {
+    log.Debugf("Pulling latest version of package %s", originName)
+    if err := pullImage(originName); err != nil {
+        return err
+    }
+    tmpdir, err := ioutil.TempDir("", "dockerappfork_")
+    if err != nil {
+        return err
+    }
+    defer os.RemoveAll(tmpdir)
+    log.Debugf("Extracting original app data to %s", tmpdir)
+    if err = Load(originName, tmpdir); err != nil {
+        return err
+    }
+
+    // create app dir in output-dir
+    namespace, name := splitPackageName(forkName)
+    appPath := filepath.Join(outputDir, internal.DirNameFromAppName(name))
+    os.MkdirAll(appPath, 0755)
+
+    // iterate tar contents
+    imgRef, err := splitImageName(originName)
+    tarfile, err := os.Open(filepath.Join(tmpdir, internal.DirNameFromAppName(imgRef.Name)))
+    if err != nil {
+        return errors.Wrap(err, "failed to open package archive")
+    }
+    tarReader := tar.NewReader(tarfile)
+    for {
+        header, err := tarReader.Next()
+        if err == io.EOF {
+            break
+        }
+        data := make([]byte, header.Size)
+        if _, err = tarReader.Read(data); err != nil && err != io.EOF {
+            return errors.Wrap(err, "error reading tar data")
+        }
+
+        if header.Name == "metadata.yml" {
+            log.Debug("Loading app metadata")
+            data, err = updateMetadata(data, namespace, name, maintainers)
+            if err != nil {
+                return err
+            }
+        }
+
+        dest := filepath.Join(appPath, header.Name)
+        log.Debugf("Writing file at %s", dest)
+        if err = ioutil.WriteFile(dest, data, 0644); err != nil {
+            return errors.Wrap(err, "error writing output file")
+        }
+    }
+
+    return nil
+}
+
+func updateMetadata(raw []byte, namespace, name string, maintainers []string) ([]byte, error) {
+    // retrieve original metadata (maintainer/app name/app tag)
+    var yamlMeta []byte
+    meta, err := loadMetadata(raw)
+    if err != nil {
+        return yamlMeta, err
+    }
+    // insert retrieved data in fork history section
+    log.Debug("Generating fork data")
+    forkData := types.ParentMetadata{
```

Comment:
We may want to extract that in a function (a `AppMetadata` _constructor_) — especially if we use the same struct (i.e. no `ParentMetadata`). Something like

```go
meta := types.MetadataFrom(origMeta, WithMaintainers(…), …)
```

Reply:
done!